# Deep Learning Fundamentals 7 - Into the World of Tensorflow/Keras 1

In the seventh notebook, I will talk about how to use the techniques we covered in the previous three notebooks by using Tensorflow and Keras. Let's get started.

In [1]:
import sys
import sklearn
import tensorflow as tf
from tensorflow import keras
import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt

# Exploring Initializers and Activation Functions 

Previously, we implemented initializers and activation functions from scratch. Now we will use them with Keras for practice, let's get started with loading our beloved dataset, Fashion MNIST.

In [2]:
(X_train, y_train_full), (X_test, y_test) = keras.datasets.fashion_mnist.load_data()
X_train_full = X_train / 255.0
X_test = X_test / 255.0
X_valid, X_train = X_train_full[:5000], X_train_full[5000:]
y_valid, y_train = y_train_full[:5000], y_train_full[5000:]

How many initializers are implemented in Keras ?

In [3]:
[initializers for initializers in dir(keras.initializers) if not initializers.startswith("_")]

['Constant',
 'GlorotNormal',
 'GlorotUniform',
 'HeNormal',
 'HeUniform',
 'Identity',
 'Initializer',
 'LecunNormal',
 'LecunUniform',
 'Ones',
 'Orthogonal',
 'RandomNormal',
 'RandomUniform',
 'TruncatedNormal',
 'VarianceScaling',
 'Zeros',
 'constant',
 'deserialize',
 'get',
 'glorot_normal',
 'glorot_uniform',
 'he_normal',
 'he_uniform',
 'identity',
 'lecun_normal',
 'lecun_uniform',
 'ones',
 'orthogonal',
 'random_normal',
 'random_uniform',
 'serialize',
 'truncated_normal',
 'variance_scaling',
 'zeros']

By default, Keras uses Glorot initialization with a uniform distribution. Let's train a network usign ReLU and default initialization then also talk about some methods that we previously used but didn't talk about much.

In [4]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.Dense(300),
    keras.layers.ReLU(),
    keras.layers.Dense(100),
    keras.layers.ReLU(),
    keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.SGD(lr=1e-2),
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))

Epoch 1/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.7044 - accuracy: 0.7687 - val_loss: 0.5176 - val_accuracy: 0.8196
Epoch 2/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.4817 - accuracy: 0.8320 - val_loss: 0.4689 - val_accuracy: 0.8322
Epoch 3/10
1719/1719 [==============================] - 3s 2ms/step - loss: 0.4381 - accuracy: 0.8460 - val_loss: 0.4297 - val_accuracy: 0.8478
Epoch 4/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.4111 - accuracy: 0.8548 - val_loss: 0.4054 - val_accuracy: 0.8594
Epoch 5/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.3913 - accuracy: 0.8625 - val_loss: 0.3782 - val_accuracy: 0.8702
Epoch 6/10
1719/1719 [==============================] - 3s 2ms/step - loss: 0.3765 - accuracy: 0.8661 - val_loss: 0.3896 - val_accuracy: 0.8678
Epoch 7/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.3624 - accuracy: 0.8721 - val_loss: 0.3762 - val_accuracy:

In the third notebook, I used Keras for training some basic networks but didn't talk much about layers, and main methods that we use such as `compile()` and `fit()`  I think now we covered more enough building blocks of deep learning to talk about these things. 

In the code above, we first defined the layers. This neural network is composed of dense layers (or a fully connected network). These layers are appropriate for 2-rank tensors of shape (samples, features). On the other hand, for tensors for different ranks we need to use different models, for instance, we can use LSTM layers, RNN layers, or 1D-CNN layers for sequence data which is stored in 3-rank tensors of shape (samples, timesteps, features). Moreover, Image data is stored in 4-rank tensors and generally processed by 2D convolution layers (Conv2D). Secondly, we use the `compile()` method and pass the optimizer, loss, and metric. I passed two of them as strings and one of them as an object (that's because I wanted to tune `learning_rate` argument), however, all of them got converted to  Python objects in the end. For instance, `loss="sparse_categorical_crossentropy"` becomes `loss=tf.keras.losses.SparseCategoricalCrossentropy` once we run the code. Lastly, I call `fit()` method which implements the training loop itself.


Additionally, there are two important methods that we frequently use 
1. `evaluate()`: This method is used when we want to compute the loss and metrics after training. If the model doesn't have a metric then only the loss gets returned.
2. `predict()`: This method is used for inference. The data we pass will be itareted over the data and in return we will get a Numpy array of predictions.

Now let's go on talking about initializations and see if He initialization will improve accuracy.

In [5]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.Dense(300, kernel_initializer="HeUniform"),
    keras.layers.ReLU(),
    keras.layers.Dense(100, kernel_initializer="HeUniform"),
    keras.layers.ReLU(),
    keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.SGD(lr=1e-2),
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))

Epoch 1/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.6898 - accuracy: 0.7735 - val_loss: 0.4971 - val_accuracy: 0.8284
Epoch 2/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.4789 - accuracy: 0.8324 - val_loss: 0.4276 - val_accuracy: 0.8616
Epoch 3/10
1719/1719 [==============================] - 3s 2ms/step - loss: 0.4337 - accuracy: 0.8486 - val_loss: 0.4115 - val_accuracy: 0.8574
Epoch 4/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.4071 - accuracy: 0.8572 - val_loss: 0.3937 - val_accuracy: 0.8672
Epoch 5/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.3878 - accuracy: 0.8647 - val_loss: 0.3825 - val_accuracy: 0.8718
Epoch 6/10
1719/1719 [==============================] - 3s 2ms/step - loss: 0.3713 - accuracy: 0.8690 - val_loss: 0.3692 - val_accuracy: 0.8746
Epoch 7/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.3592 - accuracy: 0.8734 - val_loss: 0.3679 - val_accuracy:

Looks like both initilization techniques work almost equally well in this case. How would the performance change if I used another activation function ? Let's use LeakyRelu with HeUniform initialization.

In [6]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.Dense(300, kernel_initializer="HeUniform"),
    keras.layers.LeakyReLU(),
    keras.layers.Dense(100, kernel_initializer="HeUniform"),
    keras.layers.LeakyReLU(),
    keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.SGD(lr=1e-2),
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))

Epoch 1/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.6862 - accuracy: 0.7712 - val_loss: 0.4970 - val_accuracy: 0.8360
Epoch 2/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.4866 - accuracy: 0.8297 - val_loss: 0.4651 - val_accuracy: 0.8368
Epoch 3/10
1719/1719 [==============================] - 3s 2ms/step - loss: 0.4453 - accuracy: 0.8442 - val_loss: 0.4161 - val_accuracy: 0.8532
Epoch 4/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.4208 - accuracy: 0.8531 - val_loss: 0.4173 - val_accuracy: 0.8540
Epoch 5/10
1719/1719 [==============================] - 3s 2ms/step - loss: 0.4041 - accuracy: 0.8572 - val_loss: 0.3925 - val_accuracy: 0.8644
Epoch 6/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.3898 - accuracy: 0.8635 - val_loss: 0.3932 - val_accuracy: 0.8582
Epoch 7/10
1719/1719 [==============================] - 3s 2ms/step - loss: 0.3768 - accuracy: 0.8680 - val_loss: 0.3831 - val_accuracy:

What about using PReLU ?

In [7]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.Dense(300, kernel_initializer="HeUniform"),
    keras.layers.PReLU(),
    keras.layers.Dense(100, kernel_initializer="HeUniform"),
    keras.layers.PReLU(),
    keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.SGD(lr=1e-2),
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))

Epoch 1/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.7040 - accuracy: 0.7659 - val_loss: 0.5021 - val_accuracy: 0.8308
Epoch 2/10
1719/1719 [==============================] - 3s 2ms/step - loss: 0.4835 - accuracy: 0.8311 - val_loss: 0.4598 - val_accuracy: 0.8408
Epoch 3/10
1719/1719 [==============================] - 3s 2ms/step - loss: 0.4380 - accuracy: 0.8457 - val_loss: 0.4095 - val_accuracy: 0.8608
Epoch 4/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.4123 - accuracy: 0.8551 - val_loss: 0.4177 - val_accuracy: 0.8566
Epoch 5/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.3913 - accuracy: 0.8621 - val_loss: 0.3989 - val_accuracy: 0.8630
Epoch 6/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.3736 - accuracy: 0.8677 - val_loss: 0.3980 - val_accuracy: 0.8626
Epoch 7/10
1719/1719 [==============================] - 3s 2ms/step - loss: 0.3596 - accuracy: 0.8716 - val_loss: 0.3912 - val_accuracy:

Let's also use HeNormal initialization.

In [8]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.Dense(300, kernel_initializer="HeNormal"),
    keras.layers.PReLU(),
    keras.layers.Dense(100, kernel_initializer="HeNormal"),
    keras.layers.PReLU(),
    keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.SGD(lr=1e-2),
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))

Epoch 1/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.6878 - accuracy: 0.7710 - val_loss: 0.4902 - val_accuracy: 0.8386
Epoch 2/10
1719/1719 [==============================] - 7s 4ms/step - loss: 0.4760 - accuracy: 0.8343 - val_loss: 0.4340 - val_accuracy: 0.8544
Epoch 3/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.4331 - accuracy: 0.8483 - val_loss: 0.4275 - val_accuracy: 0.8510
Epoch 4/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.4072 - accuracy: 0.8579 - val_loss: 0.3972 - val_accuracy: 0.8616
Epoch 5/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.3870 - accuracy: 0.8634 - val_loss: 0.3715 - val_accuracy: 0.8696
Epoch 6/10
1719/1719 [==============================] - 6s 4ms/step - loss: 0.3721 - accuracy: 0.8696 - val_loss: 0.3664 - val_accuracy: 0.8736
Epoch 7/10
1719/1719 [==============================] - 7s 4ms/step - loss: 0.3576 - accuracy: 0.8730 - val_loss: 0.3709 - val_accuracy:

Looks like PReLU and HeNormal provided the best result so far. 


We can also tune our initializer by changing the scaling factor. We previously talked about that He initialization uses $fan_{in}$ we can actually change this and use $fan_{avg}$ as well by using `keras.initializers.VarianceScaling()`

In [9]:
init = keras.initializers.VarianceScaling(scale=2., mode='fan_avg', distribution='normal')


model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.Dense(300, kernel_initializer=init),
    keras.layers.PReLU(),
    keras.layers.Dense(100, kernel_initializer=init),
    keras.layers.PReLU(),
    keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.SGD(lr=1e-2),
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))

Epoch 1/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.6727 - accuracy: 0.7755 - val_loss: 0.4817 - val_accuracy: 0.8424
Epoch 2/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.4755 - accuracy: 0.8349 - val_loss: 0.4686 - val_accuracy: 0.8336
Epoch 3/10
1719/1719 [==============================] - 5s 3ms/step - loss: 0.4310 - accuracy: 0.8495 - val_loss: 0.4056 - val_accuracy: 0.8608
Epoch 4/10
1719/1719 [==============================] - 6s 3ms/step - loss: 0.4057 - accuracy: 0.8575 - val_loss: 0.3870 - val_accuracy: 0.8700
Epoch 5/10
1719/1719 [==============================] - 4s 3ms/step - loss: 0.3844 - accuracy: 0.8660 - val_loss: 0.3795 - val_accuracy: 0.8696
Epoch 6/10
1719/1719 [==============================] - 4s 2ms/step - loss: 0.3696 - accuracy: 0.8701 - val_loss: 0.3960 - val_accuracy: 0.8628
Epoch 7/10
1719/1719 [==============================] - 6s 4ms/step - loss: 0.3557 - accuracy: 0.8757 - val_loss: 0.3560 - val_accuracy:

In all the examples above, we had almost similar results and didn't get any vanishing/exploiding gradient problem. However, we can see a different result when we are using a very dense neural network. Let's see how it happens.

## Let's Vanish/Explode these Gradients

Here, I will do something very basic. I will train two neural networks. In the first one I will add lots of layers by using a for loop. In the second one, I will train almost the same model with the exception that I will use a different activation function. You will see that changing just the activation function may improve our model extraordinarily in some cases. 

In [10]:
model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28]))
model.add(keras.layers.Dense(300, kernel_initializer='lecun_normal',activation='relu'))
for layer in range(50):
    model.add(keras.layers.Dense(100, kernel_initializer='lecun_normal',activation='relu')) # adding 50 hidden layers.
model.add(keras.layers.Dense(10, activation="softmax"))

model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.SGD(lr=1e-2),
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))

Epoch 1/10
1719/1719 [==============================] - 11s 7ms/step - loss: 2.3027 - accuracy: 0.0990 - val_loss: 2.3028 - val_accuracy: 0.0980
Epoch 2/10
1719/1719 [==============================] - 11s 7ms/step - loss: 2.3027 - accuracy: 0.0985 - val_loss: 2.3028 - val_accuracy: 0.0986
Epoch 3/10
1719/1719 [==============================] - 13s 7ms/step - loss: 2.3027 - accuracy: 0.0994 - val_loss: 2.3027 - val_accuracy: 0.0914
Epoch 4/10
1719/1719 [==============================] - 11s 7ms/step - loss: 2.3027 - accuracy: 0.0994 - val_loss: 2.3027 - val_accuracy: 0.0914
Epoch 5/10
1719/1719 [==============================] - 12s 7ms/step - loss: 2.3027 - accuracy: 0.0987 - val_loss: 2.3027 - val_accuracy: 0.0914
Epoch 6/10
1719/1719 [==============================] - 13s 7ms/step - loss: 2.3027 - accuracy: 0.0988 - val_loss: 2.3028 - val_accuracy: 0.0914
Epoch 7/10
1719/1719 [==============================] - 12s 7ms/step - loss: 2.3027 - accuracy: 0.0991 - val_loss: 2.3029 - val_ac

Previously we talked about a paper in which it is proposed that using SELU activation function and LeCun initialization will self-normalize our neural network (each layer will have the same mean and variance during training) and this will solve vanishing/exploding gradients problem. Let's use these proposed model and see if it improves our neural metwork.

In [11]:
model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28]))
model.add(keras.layers.Dense(300, kernel_initializer='lecun_normal',activation='selu'))
for layer in range(50):
    model.add(keras.layers.Dense(100, kernel_initializer='lecun_normal',activation='selu')) # adding 50 hidden layers.
model.add(keras.layers.Dense(10, activation="softmax"))

model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.SGD(lr=1e-2),
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))

Epoch 1/10
1719/1719 [==============================] - 13s 8ms/step - loss: 1.5003 - accuracy: 0.4050 - val_loss: 0.8019 - val_accuracy: 0.6812
Epoch 2/10
1719/1719 [==============================] - 14s 8ms/step - loss: 0.7328 - accuracy: 0.7153 - val_loss: 0.8693 - val_accuracy: 0.6672
Epoch 3/10
1719/1719 [==============================] - 13s 7ms/step - loss: 0.6804 - accuracy: 0.7614 - val_loss: 0.5321 - val_accuracy: 0.8148
Epoch 4/10
1719/1719 [==============================] - 14s 8ms/step - loss: 0.5514 - accuracy: 0.8084 - val_loss: 0.4932 - val_accuracy: 0.8326
Epoch 5/10
1719/1719 [==============================] - 13s 8ms/step - loss: 0.4874 - accuracy: 0.8314 - val_loss: 0.4818 - val_accuracy: 0.8234
Epoch 6/10
1719/1719 [==============================] - 14s 8ms/step - loss: 0.4724 - accuracy: 0.8355 - val_loss: 0.4912 - val_accuracy: 0.8382
Epoch 7/10
1719/1719 [==============================] - 13s 8ms/step - loss: 0.4438 - accuracy: 0.8448 - val_loss: 0.4243 - val_ac

**It worked !!** In the first model, we suffered from vanishing/exploiding gradient problem, however, this wasn't the case in the second example. But there are still some point that we need to be careful about which are stated in [Géron, A. (2019)](https://www.oreilly.com/library/view/hands-on-machine-learning/9781492032632/).
* The self-normalizing property of the SELU activation function is easily broken: you cannot use ℓ<sub>1</sub> or ℓ<sub>2</sub> regularization, regular dropout, max-norm, skip connections or other non-sequential topologies (so recurrent neural networks won't self-normalize). However, in practice it works quite well with sequential CNNs. If you break self-normalization, SELU will not necessarily outperform other activation functions.

However, if we don't use any technique that break the self-normalizing property. The function preserves this property for even very big neural networks. Moreover, by default, the hyperparameters of SELU  are tuned in a way that the mean output of each neuron remains close to 0, and the standard deviation remains close to 1.

The below suggestion is also taken from [the book](https://www.oreilly.com/library/view/hands-on-machine-learning/9781492032632/):


**Which activation function should you use for the hidden layers of your deep neural networks?** 
* Although your mileage will vary, in general SELU > ELU > leaky ReLU (and its variants) > ReLU > tanh > logistic. If the network’s architecture prevents it from self-normalizing, then ELU may perform better than SELU (since SELU is not smooth at z = 0). If you care a lot about runtime latency, then you may prefer leaky ReLU. If you don’t want to tweak yet another hyperparameter, you may use the default α values used by Keras (e.g., 0.3 for leaky ReLU). If you have spare time and computing power, you can use cross-validation to evaluate other activation functions, such as RReLU if your network is overfitting or PReLU if you have a huge training set. That said, because ReLU is the most used activation function (by far), many libraries and hardware accelerators provide ReLU-specific optimizations; therefore, if speed is your priority, ReLU might still be the best choice - [Géron, A. (2019)](https://www.oreilly.com/library/view/hands-on-machine-learning/9781492032632/).

# Batch Normalization

We previously talked about that if we add BN as the very first layer of our neural network we generally don't need to standardize our training set. Moreover, sometimes applying BN before the activation function works better (try both). Also layer before a BN layer doesn't need to have bias terms, therefore, it would be appropriate to use `use_bias=False` to avoid wasting parameters (This is actually because that in the mean subtraction step of batch normalization, the bias term will be canceled out since we are adding it to all neurons).

In [12]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(300, kernel_initializer="HeUniform"),
    keras.layers.PReLU(),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(100, kernel_initializer="HeUniform"),
    keras.layers.PReLU(),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(10, activation="softmax")
])

model.compile(loss="sparse_categorical_crossentropy",
              optimizer=keras.optimizers.SGD(lr=1e-2),
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid),batch_size=64)

Epoch 1/10
860/860 [==============================] - 5s 6ms/step - loss: 0.6066 - accuracy: 0.7929 - val_loss: 0.4351 - val_accuracy: 0.8438
Epoch 2/10
860/860 [==============================] - 6s 7ms/step - loss: 0.4230 - accuracy: 0.8520 - val_loss: 0.3856 - val_accuracy: 0.8582
Epoch 3/10
860/860 [==============================] - 5s 6ms/step - loss: 0.3792 - accuracy: 0.8646 - val_loss: 0.3662 - val_accuracy: 0.8666
Epoch 4/10
860/860 [==============================] - 6s 7ms/step - loss: 0.3518 - accuracy: 0.8746 - val_loss: 0.3599 - val_accuracy: 0.8738
Epoch 5/10
860/860 [==============================] - 6s 7ms/step - loss: 0.3313 - accuracy: 0.8816 - val_loss: 0.3438 - val_accuracy: 0.8772
Epoch 6/10
860/860 [==============================] - 5s 5ms/step - loss: 0.3133 - accuracy: 0.8865 - val_loss: 0.3389 - val_accuracy: 0.8782
Epoch 7/10
860/860 [==============================] - 5s 5ms/step - loss: 0.2986 - accuracy: 0.8922 - val_loss: 0.3341 - val_accuracy: 0.8820
Epoch 

Let's see the model summary.

In [13]:
model.summary()

Model: "sequential_8"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_8 (Flatten)          (None, 784)               0         
_________________________________________________________________
batch_normalization (BatchNo (None, 784)               3136      
_________________________________________________________________
dense_122 (Dense)            (None, 300)               235500    
_________________________________________________________________
p_re_lu_6 (PReLU)            (None, 300)               300       
_________________________________________________________________
batch_normalization_1 (Batch (None, 300)               1200      
_________________________________________________________________
dense_123 (Dense)            (None, 100)               30100     
_________________________________________________________________
p_re_lu_7 (PReLU)            (None, 100)              

Let's also look at the parameters BN layer added to our model.

In [14]:
[(var.name, var.trainable) for var in model.layers[4].variables]

[('batch_normalization_1/gamma:0', True),
 ('batch_normalization_1/beta:0', True),
 ('batch_normalization_1/moving_mean:0', False),
 ('batch_normalization_1/moving_variance:0', False)]

Moving mean and moving variance are non-trainable.

Moreover, the BatchNormalization class has some hyperparameters we can tune such as the momentum (the hyperparameter that is used to update the exponential moving averages given a new value v (i.e., a new vector of input means or standard deviations computed over the current batch). A good momentum should be generally close to 1. The other hyperparameter is an axis which determines the axis that should be normalized and it is -1 by default which corresponds that the last axis will be normalized.


The below paragraph is taken from [the book](https://www.oreilly.com/library/view/hands-on-machine-learning/9781492032632/)
* When the input batch is 2D (i.e., the batch shape is [batch size, features]), this means that each input feature will be normalized based on the mean and standard deviation computed across all the instances in the batch. For example, the first BN layer in the previous code example will independently normalize (and rescale and shift) each of the 784 input features. If we move the first BN layer before the Flatten layer, then the input batches will be 3D, with shape [batch size, height, width]; therefore, the BN layer will compute 28 means and 28 standard deviations (1 per column of pixels, computed across all instances in the batch and across all rows in the column), and it will normalize all pixels in a given column using the same mean and standard deviation. There will also be just 28 scale parameters and 28 shift parameters. If instead you still want to treat each of the 784 pixels independently, then you should set axis=[1, 2]. Notice that the BN layer does not perform the same computation during training and after training: it uses batch statistics during training and the “final” statistics after training (i.e., the final values of the moving averages).

A good article for PReLU [link](https://github.com/christianversloot/machine-learning-articles/blob/main/how-to-use-prelu-with-keras.md)

Important paper : [Fixup Initialization: Residual Learning Without Normalization](https://arxiv.org/abs/1901.09321)

# Optimization

Let's use the model that gave us the best scores with different optimization functions to see whether we can improve it further.

In [15]:
def Optimizer_tryout(selected):
    optimizer = selected
    model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(300, kernel_initializer="HeUniform"),
    keras.layers.PReLU(),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(100, kernel_initializer="HeUniform"),
    keras.layers.PReLU(),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(10, activation="softmax")
])

    model.compile(loss="sparse_categorical_crossentropy",
              optimizer=optimizer,
              metrics=["accuracy"])

    history = model.fit(X_train, y_train, epochs=10,
                      validation_data=(X_valid, y_valid))

## Momentum optimization

In [16]:
Optimizer_tryout(keras.optimizers.SGD(lr=0.001, momentum=0.9))

Epoch 1/10
1719/1719 [==============================] - 7s 4ms/step - loss: 0.5652 - accuracy: 0.8026 - val_loss: 0.3900 - val_accuracy: 0.8670
Epoch 2/10
1719/1719 [==============================] - 6s 3ms/step - loss: 0.4089 - accuracy: 0.8535 - val_loss: 0.3573 - val_accuracy: 0.8722
Epoch 3/10
1719/1719 [==============================] - 8s 4ms/step - loss: 0.3674 - accuracy: 0.8677 - val_loss: 0.3354 - val_accuracy: 0.8822
Epoch 4/10
1719/1719 [==============================] - 7s 4ms/step - loss: 0.3407 - accuracy: 0.8763 - val_loss: 0.3236 - val_accuracy: 0.8884
Epoch 5/10
1719/1719 [==============================] - 7s 4ms/step - loss: 0.3205 - accuracy: 0.8851 - val_loss: 0.3163 - val_accuracy: 0.8910
Epoch 6/10
1719/1719 [==============================] - 7s 4ms/step - loss: 0.3032 - accuracy: 0.8903 - val_loss: 0.3166 - val_accuracy: 0.8908
Epoch 7/10
1719/1719 [==============================] - 6s 4ms/step - loss: 0.2884 - accuracy: 0.8956 - val_loss: 0.3093 - val_accuracy:

## Nesterov Accelerated Gradient

In [17]:
Optimizer_tryout(keras.optimizers.SGD(lr=0.001, momentum=0.9, nesterov=True))

Epoch 1/10
1719/1719 [==============================] - 7s 4ms/step - loss: 0.5726 - accuracy: 0.7994 - val_loss: 0.3951 - val_accuracy: 0.8658
Epoch 2/10
1719/1719 [==============================] - 7s 4ms/step - loss: 0.4106 - accuracy: 0.8529 - val_loss: 0.3604 - val_accuracy: 0.8732
Epoch 3/10
1719/1719 [==============================] - 7s 4ms/step - loss: 0.3665 - accuracy: 0.8684 - val_loss: 0.3418 - val_accuracy: 0.8800
Epoch 4/10
1719/1719 [==============================] - 7s 4ms/step - loss: 0.3424 - accuracy: 0.8774 - val_loss: 0.3244 - val_accuracy: 0.8846
Epoch 5/10
1719/1719 [==============================] - 8s 5ms/step - loss: 0.3191 - accuracy: 0.8851 - val_loss: 0.3204 - val_accuracy: 0.8852
Epoch 6/10
1719/1719 [==============================] - 8s 5ms/step - loss: 0.3038 - accuracy: 0.8907 - val_loss: 0.3186 - val_accuracy: 0.8900
Epoch 7/10
1719/1719 [==============================] - 7s 4ms/step - loss: 0.2883 - accuracy: 0.8956 - val_loss: 0.3082 - val_accuracy:

## AdaGrad

In [18]:
Optimizer_tryout(keras.optimizers.Adagrad(lr=0.001))

Epoch 1/10
1719/1719 [==============================] - 11s 6ms/step - loss: 0.7339 - accuracy: 0.7568 - val_loss: 0.5051 - val_accuracy: 0.8318
Epoch 2/10
1719/1719 [==============================] - 9s 5ms/step - loss: 0.5321 - accuracy: 0.8175 - val_loss: 0.4483 - val_accuracy: 0.8484
Epoch 3/10
1719/1719 [==============================] - 8s 4ms/step - loss: 0.4868 - accuracy: 0.8317 - val_loss: 0.4208 - val_accuracy: 0.8556
Epoch 4/10
1719/1719 [==============================] - 9s 5ms/step - loss: 0.4606 - accuracy: 0.8415 - val_loss: 0.4010 - val_accuracy: 0.8618
Epoch 5/10
1719/1719 [==============================] - 9s 5ms/step - loss: 0.4403 - accuracy: 0.8480 - val_loss: 0.3928 - val_accuracy: 0.8648
Epoch 6/10
1719/1719 [==============================] - 9s 5ms/step - loss: 0.4279 - accuracy: 0.8523 - val_loss: 0.3836 - val_accuracy: 0.8680
Epoch 7/10
1719/1719 [==============================] - 12s 7ms/step - loss: 0.4157 - accuracy: 0.8550 - val_loss: 0.3761 - val_accurac

## RMSProp

In [19]:
Optimizer_tryout(keras.optimizers.RMSprop(lr=0.001, rho=0.9))

Epoch 1/10
1719/1719 [==============================] - 18s 10ms/step - loss: 0.4721 - accuracy: 0.8307 - val_loss: 0.3514 - val_accuracy: 0.8730
Epoch 2/10
1719/1719 [==============================] - 18s 10ms/step - loss: 0.3623 - accuracy: 0.8690 - val_loss: 0.3202 - val_accuracy: 0.8862
Epoch 3/10
1719/1719 [==============================] - 15s 9ms/step - loss: 0.3228 - accuracy: 0.8820 - val_loss: 0.3132 - val_accuracy: 0.8886
Epoch 4/10
1719/1719 [==============================] - 15s 9ms/step - loss: 0.2969 - accuracy: 0.8910 - val_loss: 0.2999 - val_accuracy: 0.8922
Epoch 5/10
1719/1719 [==============================] - 13s 8ms/step - loss: 0.2777 - accuracy: 0.8981 - val_loss: 0.3019 - val_accuracy: 0.8952
Epoch 6/10
1719/1719 [==============================] - 13s 8ms/step - loss: 0.2582 - accuracy: 0.9049 - val_loss: 0.3106 - val_accuracy: 0.8926
Epoch 7/10
1719/1719 [==============================] - 14s 8ms/step - loss: 0.2444 - accuracy: 0.9102 - val_loss: 0.2904 - val_

## Adam Optimization

In [20]:
Optimizer_tryout(keras.optimizers.Adam(lr=0.001, beta_1=0.9, beta_2=0.999))

Epoch 1/10
1719/1719 [==============================] - 8s 4ms/step - loss: 0.4721 - accuracy: 0.8302 - val_loss: 0.3586 - val_accuracy: 0.8704
Epoch 2/10
1719/1719 [==============================] - 11s 7ms/step - loss: 0.3635 - accuracy: 0.8667 - val_loss: 0.3120 - val_accuracy: 0.8816
Epoch 3/10
1719/1719 [==============================] - 10s 6ms/step - loss: 0.3230 - accuracy: 0.8811 - val_loss: 0.2984 - val_accuracy: 0.8890
Epoch 4/10
1719/1719 [==============================] - 9s 5ms/step - loss: 0.2968 - accuracy: 0.8901 - val_loss: 0.3109 - val_accuracy: 0.8892
Epoch 5/10
1719/1719 [==============================] - 7s 4ms/step - loss: 0.2815 - accuracy: 0.8952 - val_loss: 0.3128 - val_accuracy: 0.8832
Epoch 6/10
1719/1719 [==============================] - 13s 8ms/step - loss: 0.2598 - accuracy: 0.9032 - val_loss: 0.3171 - val_accuracy: 0.8878
Epoch 7/10
1719/1719 [==============================] - 11s 7ms/step - loss: 0.2446 - accuracy: 0.9078 - val_loss: 0.3008 - val_accur

# Adam Types

1. Adamax: This version of Adam is proposed in the same paper that Adam is proposed. The Adam optimization algorithm scales down the parameter updates by $L2$ norm of the time-decayed gradient, on the other hand, Adamax uses $L_{max}$ norm and scales the parameter updates by the max of the time-decayed gradients. In theory this adjusment makes Adamax more stable but it is up to the dataset in practice.


2. Nadam: Nadam algorithm uses Nesterov trick with Adam optimization which makes it converge slightly faster than Adam. In the [paper](https://cs229.stanford.edu/proj2015/054_report.pdf) that algorithm is proposed, it is also reported that this algorithm generally outperform Adam but at the same time sometimes outperformed by RMSProp.

## Adamax Optimization

In [21]:
Optimizer_tryout(keras.optimizers.Adamax(lr=0.001, beta_1=0.9, beta_2=0.999))

Epoch 1/10
1719/1719 [==============================] - 12s 7ms/step - loss: 0.4845 - accuracy: 0.8279 - val_loss: 0.3641 - val_accuracy: 0.8718
Epoch 2/10
1719/1719 [==============================] - 14s 8ms/step - loss: 0.3635 - accuracy: 0.8686 - val_loss: 0.3253 - val_accuracy: 0.8820
Epoch 3/10
1719/1719 [==============================] - 12s 7ms/step - loss: 0.3226 - accuracy: 0.8824 - val_loss: 0.3111 - val_accuracy: 0.8844
Epoch 4/10
1719/1719 [==============================] - 11s 6ms/step - loss: 0.2930 - accuracy: 0.8928 - val_loss: 0.3053 - val_accuracy: 0.8904
Epoch 5/10
1719/1719 [==============================] - 12s 7ms/step - loss: 0.2716 - accuracy: 0.9013 - val_loss: 0.3016 - val_accuracy: 0.8904
Epoch 6/10
1719/1719 [==============================] - 9s 5ms/step - loss: 0.2519 - accuracy: 0.9068 - val_loss: 0.2903 - val_accuracy: 0.8944
Epoch 7/10
1719/1719 [==============================] - 10s 6ms/step - loss: 0.2382 - accuracy: 0.9123 - val_loss: 0.2937 - val_acc

## Nadam Optimization

In [22]:
Optimizer_tryout(keras.optimizers.Nadam(lr=0.001, beta_1=0.9, beta_2=0.999))

Epoch 1/10
1719/1719 [==============================] - 16s 9ms/step - loss: 0.4681 - accuracy: 0.8333 - val_loss: 0.3441 - val_accuracy: 0.8780
Epoch 2/10
1719/1719 [==============================] - 17s 10ms/step - loss: 0.3574 - accuracy: 0.8683 - val_loss: 0.3280 - val_accuracy: 0.8818
Epoch 3/10
1719/1719 [==============================] - 20s 11ms/step - loss: 0.3219 - accuracy: 0.8815 - val_loss: 0.3115 - val_accuracy: 0.8894
Epoch 4/10
1719/1719 [==============================] - 17s 10ms/step - loss: 0.2955 - accuracy: 0.8899 - val_loss: 0.2968 - val_accuracy: 0.8944
Epoch 5/10
1719/1719 [==============================] - 14s 8ms/step - loss: 0.2743 - accuracy: 0.8962 - val_loss: 0.3010 - val_accuracy: 0.8894
Epoch 6/10
1719/1719 [==============================] - 20s 11ms/step - loss: 0.2584 - accuracy: 0.9025 - val_loss: 0.2969 - val_accuracy: 0.8908
Epoch 7/10
1719/1719 [==============================] - 16s 9ms/step - loss: 0.2446 - accuracy: 0.9081 - val_loss: 0.2858 - va

Also see these versions of Adam
1. [Adagrad](https://keras.io/api/optimizers/adagrad/)
2. [Adadelta](https://keras.io/api/optimizers/adadelta/)

**Notes from [the book](https://www.oreilly.com/library/view/hands-on-machine-learning/9781492032632/):**

* Adaptive optimization methods (including RMSProp, Adam, and Nadam optimization) are often great, converging fast to a good solution. However, a 2017 [paper](https://arxiv.org/abs/1705.08292) by Ashia C. Wilson et al. showed that they can lead to solutions that generalize poorly on some datasets. So when you are disappointed by your model’s performance, try using plain Nesterov Accelerated Gradient instead: your dataset may just be allergic to adaptive gradients. Also check out the latest research, because it’s moving fast.

# Regularization

Let's add L2-norm to our neural network and see how it affects the performance.

In [23]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(300, use_bias=False,kernel_regularizer='l2'),
    keras.layers.BatchNormalization(),
    keras.layers.Activation("relu"),
    keras.layers.Dense(100, use_bias=False,kernel_regularizer='l2'),
    keras.layers.BatchNormalization(),
    keras.layers.Activation("relu"),
    keras.layers.Dense(10, activation="softmax",kernel_regularizer='l2')
])

model.compile(loss="sparse_categorical_crossentropy",
              optimizer="nadam",
              metrics=["accuracy"])

history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))

Epoch 1/10
1719/1719 [==============================] - 20s 11ms/step - loss: 1.2700 - accuracy: 0.8011 - val_loss: 0.7218 - val_accuracy: 0.8214
Epoch 2/10
1719/1719 [==============================] - 19s 11ms/step - loss: 0.7316 - accuracy: 0.8155 - val_loss: 0.6579 - val_accuracy: 0.8376
Epoch 3/10
1719/1719 [==============================] - 20s 12ms/step - loss: 0.6688 - accuracy: 0.8267 - val_loss: 0.5892 - val_accuracy: 0.8548
Epoch 4/10
1719/1719 [==============================] - 19s 11ms/step - loss: 0.6204 - accuracy: 0.8346 - val_loss: 0.5404 - val_accuracy: 0.8578
Epoch 5/10
1719/1719 [==============================] - 18s 10ms/step - loss: 0.5950 - accuracy: 0.8375 - val_loss: 0.5519 - val_accuracy: 0.8530
Epoch 6/10
1719/1719 [==============================] - 21s 12ms/step - loss: 0.5725 - accuracy: 0.8412 - val_loss: 0.5199 - val_accuracy: 0.8594
Epoch 7/10
1719/1719 [==============================] - 21s 12ms/step - loss: 0.5636 - accuracy: 0.8423 - val_loss: 0.5265 -

Now I will use L1-norm but I will pass the argument as Python Object because I want to tune the penality.

In [24]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(300, activation="elu",
                       kernel_initializer="he_normal",
                       kernel_regularizer=keras.regularizers.l1(0.01)),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(100, activation="elu",
                       kernel_initializer="he_normal",
                       kernel_regularizer=keras.regularizers.l1(0.01)),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(10, activation="softmax",
                       kernel_regularizer=keras.regularizers.l1(0.01))
])
model.compile(loss="sparse_categorical_crossentropy", optimizer="nadam", metrics=["accuracy"])
history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))

Epoch 1/10
1719/1719 [==============================] - 18s 10ms/step - loss: 6.1403 - accuracy: 0.7226 - val_loss: 1.8012 - val_accuracy: 0.8022
Epoch 2/10
1719/1719 [==============================] - 16s 9ms/step - loss: 1.7472 - accuracy: 0.7788 - val_loss: 1.4452 - val_accuracy: 0.8248
Epoch 3/10
1719/1719 [==============================] - 27s 16ms/step - loss: 1.5018 - accuracy: 0.7910 - val_loss: 1.3863 - val_accuracy: 0.8002
Epoch 4/10
1719/1719 [==============================] - 23s 13ms/step - loss: 1.3621 - accuracy: 0.7981 - val_loss: 1.2265 - val_accuracy: 0.8300
Epoch 5/10
1719/1719 [==============================] - 18s 10ms/step - loss: 1.2634 - accuracy: 0.8017 - val_loss: 1.1502 - val_accuracy: 0.8334
Epoch 6/10
1719/1719 [==============================] - 16s 9ms/step - loss: 1.1938 - accuracy: 0.8060 - val_loss: 1.1017 - val_accuracy: 0.8332
Epoch 7/10
1719/1719 [==============================] - 15s 9ms/step - loss: 1.1490 - accuracy: 0.8057 - val_loss: 1.0920 - va

Instead of defining the activation function, initializer and regularizer in each layer we can use `functools.partial()` function to create a wrapper and call it with arguments we defined. I will create the wrapper with elastic net term.

In [25]:
from functools import partial

RegularizedDense = partial(keras.layers.Dense,
                           activation="selu",
                           kernel_initializer="he_normal",
                           kernel_regularizer=keras.regularizers.l1_l2(l1=0.01,l2=0.01))

model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.BatchNormalization(),
    RegularizedDense(300),
    keras.layers.BatchNormalization(),
    RegularizedDense(100),
    keras.layers.BatchNormalization(),
    RegularizedDense(10, activation="softmax")
])
model.compile(loss="sparse_categorical_crossentropy", optimizer="nadam", metrics=["accuracy"])
history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))

Epoch 1/10
1719/1719 [==============================] - 22s 13ms/step - loss: 6.4042 - accuracy: 0.7169 - val_loss: 1.8937 - val_accuracy: 0.7830
Epoch 2/10
1719/1719 [==============================] - 21s 12ms/step - loss: 1.8271 - accuracy: 0.7726 - val_loss: 1.7095 - val_accuracy: 0.7910
Epoch 3/10
1719/1719 [==============================] - 20s 11ms/step - loss: 1.5268 - accuracy: 0.7892 - val_loss: 1.3467 - val_accuracy: 0.8176
Epoch 4/10
1719/1719 [==============================] - 19s 11ms/step - loss: 1.3808 - accuracy: 0.7929 - val_loss: 1.3027 - val_accuracy: 0.8060
Epoch 5/10
1719/1719 [==============================] - 19s 11ms/step - loss: 1.2872 - accuracy: 0.7968 - val_loss: 1.1993 - val_accuracy: 0.8222
Epoch 6/10
1719/1719 [==============================] - 19s 11ms/step - loss: 1.2124 - accuracy: 0.8000 - val_loss: 1.1425 - val_accuracy: 0.8210
Epoch 7/10
1719/1719 [==============================] - 20s 12ms/step - loss: 1.1643 - accuracy: 0.8017 - val_loss: 1.0901 -

## Dropout

Now let's implement Dropout regularization.

In [26]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.Dropout(rate=0.25),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(300, activation="elu", kernel_initializer="he_normal"),
    keras.layers.Dropout(rate=0.25),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(100, activation="elu", kernel_initializer="he_normal"),
    keras.layers.Dropout(rate=0.25),
    keras.layers.BatchNormalization(),
    keras.layers.Dense(10, activation="softmax")
])
model.compile(loss="sparse_categorical_crossentropy", optimizer="nadam", metrics=["accuracy"])
history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))

Epoch 1/10
1719/1719 [==============================] - 17s 10ms/step - loss: 0.6017 - accuracy: 0.7844 - val_loss: 0.3849 - val_accuracy: 0.8630
Epoch 2/10
1719/1719 [==============================] - 16s 9ms/step - loss: 0.4751 - accuracy: 0.8258 - val_loss: 0.3612 - val_accuracy: 0.8694
Epoch 3/10
1719/1719 [==============================] - 23s 13ms/step - loss: 0.4409 - accuracy: 0.8391 - val_loss: 0.3324 - val_accuracy: 0.8724
Epoch 4/10
1719/1719 [==============================] - 21s 12ms/step - loss: 0.4189 - accuracy: 0.8452 - val_loss: 0.3276 - val_accuracy: 0.8832
Epoch 5/10
1719/1719 [==============================] - 23s 13ms/step - loss: 0.4030 - accuracy: 0.8510 - val_loss: 0.3076 - val_accuracy: 0.8824
Epoch 6/10
1719/1719 [==============================] - 23s 13ms/step - loss: 0.3943 - accuracy: 0.8555 - val_loss: 0.3058 - val_accuracy: 0.8832
Epoch 7/10
1719/1719 [==============================] - 20s 11ms/step - loss: 0.3856 - accuracy: 0.8561 - val_loss: 0.3056 - 

I used BatchNorm after Dropout which is because using it before may cause information leakage.

**Notes from [the book](https://www.oreilly.com/library/view/hands-on-machine-learning/9781492032632/):**

* Since dropout is only active during training, comparing the training loss and the validation loss can be misleading. In particular, a model may be overfitting the training set and yet have similar training and validation losses. So make sure to evaluate the training loss without dropout (e.g., after training).

## Alpha Dropout

One problem of regular dropout is that it breaks self-normalization feature of SELU + Lecun_normalization combination. To cope with that, we can use alpha dropout which preserves the mean and standard deviation of its inputs.

In [27]:
model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    keras.layers.AlphaDropout(rate=0.2),
    keras.layers.Dense(300, activation="selu", kernel_initializer="lecun_normal"),
    keras.layers.AlphaDropout(rate=0.2),
    keras.layers.Dense(100, activation="selu", kernel_initializer="lecun_normal"),
    keras.layers.AlphaDropout(rate=0.2),
    keras.layers.Dense(10, activation="softmax")
])
optimizer = keras.optimizers.SGD(lr=0.01, momentum=0.9, nesterov=True)
model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=["accuracy"])
history = model.fit(X_train, y_train, epochs=10,
                    validation_data=(X_valid, y_valid))

Epoch 1/10
1719/1719 [==============================] - 7s 4ms/step - loss: 1.0597 - accuracy: 0.6018 - val_loss: 0.9336 - val_accuracy: 0.7078
Epoch 2/10
1719/1719 [==============================] - 6s 4ms/step - loss: 0.9141 - accuracy: 0.6573 - val_loss: 0.8956 - val_accuracy: 0.7246
Epoch 3/10
1719/1719 [==============================] - 7s 4ms/step - loss: 0.8861 - accuracy: 0.6666 - val_loss: 0.9292 - val_accuracy: 0.6994
Epoch 4/10
1719/1719 [==============================] - 6s 4ms/step - loss: 0.8689 - accuracy: 0.6739 - val_loss: 1.0638 - val_accuracy: 0.6956
Epoch 5/10
1719/1719 [==============================] - 6s 4ms/step - loss: 0.8586 - accuracy: 0.6782 - val_loss: 0.9549 - val_accuracy: 0.7108
Epoch 6/10
1719/1719 [==============================] - 6s 4ms/step - loss: 0.8385 - accuracy: 0.6829 - val_loss: 0.8677 - val_accuracy: 0.7532
Epoch 7/10
1719/1719 [==============================] - 6s 3ms/step - loss: 0.8334 - accuracy: 0.6857 - val_loss: 0.7209 - val_accuracy:

## Monte Carlo Dropout - A free lunch ?

An interesting technique called Monte Carlo Dropout published in a [paper](https://arxiv.org/abs/1506.02142) by Yarin Gal and Zoubin Ghahramani states that there is a good connection between dropout networks and Deep Gaussian Process (An approximate bayesian inference method). When we use dropout in each iteration, we get a slightly different neural network architecture. While doing predictions for test data we can use these different architectures and get the average prediction results which will cause our model to predict slightly better. You can read more about this algorithm in this [article](https://towardsdatascience.com/monte-carlo-dropout-7fd52f8b6571). The algorithm resembles a voting classifier.

Let's make 100 predictions over the test set (we need to set `training=True` to ensure that the Dropout layer is active) and then we stack the predictions. Thanks to dropout predictions are made by different arhitectures, in other words, we have different predictions. When we average over the first dimension (axis=0) we will get a array of shape that we would get with a single prediction. In the end, averaging over multiple predictions with dropout on gives us a Monte Carlo estimate that is generally more reliable than the result of a single prediction with dropout off.

In [28]:
y_probas = np.stack([model(X_test, training=True)
                     for sample in range(100)])
y_proba = y_probas.mean(axis=0)
y_std = y_probas.std(axis=0)

Let's look at the prediction for the first instance in the test data.

In [29]:
np.round(model.predict(X_test[:1]), 2)

array([[0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.  , 0.01, 0.  , 0.99]],
      dtype=float32)

The model says that that instance belong to class 9 with %97 probability. Let's see how this probability will change when we use Monte Carlo Dropout.

In [ ]:
np.round(y_probas[:, :1], 2)

In the end, the model still says that this instance belong to the class 9. However, it is more unsure about the class of instance. Let's average over the first dimension to get the classification result for the first instance.

In [31]:
np.round(y_proba[:1], 2)

array([[0.  , 0.  , 0.  , 0.  , 0.  , 0.3 , 0.  , 0.2 , 0.04, 0.45]],
      dtype=float32)

We can also have a look at the standard deviation of the probability estimation.

In [32]:
y_std = y_probas.std(axis=0)
np.round(y_std[:1], 2)

array([[0.02, 0.01, 0.  , 0.02, 0.  , 0.21, 0.02, 0.14, 0.08, 0.25]],
      dtype=float32)

Looks like there is a big variance in the probability estimation. It is not that important for this model but it may be very important especially while trying to build a model for medical prediction. Let's also see whether we improved the accuracy.

In [33]:
y_pred = np.argmax(y_proba, axis=1)

In [34]:
accuracy = np.sum(y_pred == y_test) / len(y_test)
accuracy

0.7792

Looks like we improved the accuracy.

*  The number of Monte Carlo samples you use (100 in this example) is a hyperparameter you can tweak. The higher it is, the more accurate the predictions and their uncertainty estimates will be. However, if you double it, inference time will also be doubled. Moreover, above a certain number of samples, you will notice little improvement. So your job is to find the right trade-off between latency and accuracy, depending on your application - [Géron, A. (2019)](https://www.oreilly.com/library/view/hands-on-machine-learning/9781492032632/)

It is not a good idea to force training mode if we have special layers like BatchNormalization layers. In cases like that we can use Subclass API and override the call() method to directly force its training argument. We can also do the same thing for Alpha Dropout and get a MCAlpha Dropout version.

In [35]:
class MCDropout(keras.layers.Dropout):
    def call(self, inputs):
        return super().call(inputs, training=True)

class MCAlphaDropout(keras.layers.AlphaDropout):
    def call(self, inputs):
        return super().call(inputs, training=True)

Let's use MCAlphaDropout

In [36]:
mc_model = keras.models.Sequential([
    MCAlphaDropout(layer.rate) if isinstance(layer, keras.layers.AlphaDropout) else layer
    for layer in model.layers
])

In [37]:
mc_model.summary()

Model: "sequential_21"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_20 (Flatten)         (None, 784)               0         
_________________________________________________________________
mc_alpha_dropout (MCAlphaDro (None, 784)               0         
_________________________________________________________________
dense_158 (Dense)            (None, 300)               235500    
_________________________________________________________________
mc_alpha_dropout_1 (MCAlphaD (None, 300)               0         
_________________________________________________________________
dense_159 (Dense)            (None, 100)               30100     
_________________________________________________________________
mc_alpha_dropout_2 (MCAlphaD (None, 100)               0         
_________________________________________________________________
dense_160 (Dense)            (None, 10)              

In [38]:
optimizer = keras.optimizers.SGD(lr=0.01, momentum=0.9, nesterov=True)
mc_model.compile(loss="sparse_categorical_crossentropy", optimizer=optimizer, metrics=["accuracy"])

In [39]:
mc_model.set_weights(model.get_weights())

In [40]:
np.round(np.mean([mc_model.predict(X_test[:1]) for sample in range(100)], axis=0), 3)

array([[0.002, 0.   , 0.001, 0.001, 0.002, 0.339, 0.003, 0.207, 0.025,
        0.42 ]], dtype=float32)

In [41]:
y_probas = np.stack([mc_model(X_test) for sample in range(100)])
y_proba = y_probas.mean(axis=0)

In [42]:
y_pred = np.argmax(y_proba, axis=1)

In [43]:
accuracy = np.sum(y_pred == y_test) / len(y_test)
accuracy

0.7803

Another way to calculate accuracy

In [44]:
y_pred = np.argmax(np.round(np.mean([mc_model.predict(X_test) for sample in range(100)], axis=0), 3), axis=1)

In [45]:
accuracy = np.sum(y_pred == y_test) / len(y_test)
accuracy

0.7806

This MCDropout class will work with all Keras APIs, including the Sequential API. If you only care about the
Functional API or the Subclassing API, you do not have to create an MCDropout class; you can create a regular
Dropout layer and call it with `training=True`.

## Max Norm Regularization

Max-Norm Regularization or Max-Norm Constraint is a popular regularization technique in which we constrain the weights of connections for each neuron. It uses L2 term even though it does not actually add the term into the overall loss function. Instead, after each training step the weight vector is forced to have L2 norm if it is less than or equal to the hyperparameter **r**. If this condition is not satisfied, the weight vector is replaced with by the unit vector that is scaled by r. 

You can find more about Max-Norm Regularization in the following articles.

[This one is about Max-Norm Regularization](https://machinelearningjourney.com/index.php/2021/01/15/max-norm-regularization/)  

[This one contains information about all the regularization methods](https://cs231n.github.io/neural-networks-2/#reg)

In [46]:
layer = keras.layers.Dense(100, activation="selu", kernel_initializer="lecun_normal",
                           kernel_constraint=keras.constraints.max_norm(1.))

In [47]:
MaxNormDense = partial(keras.layers.Dense,
                       activation="selu", kernel_initializer="lecun_normal",
                       kernel_constraint=keras.constraints.max_norm(1.))

model = keras.models.Sequential([
    keras.layers.Flatten(input_shape=[28, 28]),
    MaxNormDense(300),
    MaxNormDense(100),
    keras.layers.Dense(10, activation="softmax")
])
model.compile(loss="sparse_categorical_crossentropy", optimizer="nadam", metrics=["accuracy"])
history = model.fit(X_train, y_train, epochs=20,
                    validation_data=(X_valid, y_valid))

Epoch 1/20
1719/1719 [==============================] - 13s 7ms/step - loss: 0.5042 - accuracy: 0.8164 - val_loss: 0.4153 - val_accuracy: 0.8446
Epoch 2/20
1719/1719 [==============================] - 15s 9ms/step - loss: 0.3976 - accuracy: 0.8552 - val_loss: 0.3892 - val_accuracy: 0.8576
Epoch 3/20
1719/1719 [==============================] - 16s 10ms/step - loss: 0.3722 - accuracy: 0.8634 - val_loss: 0.4403 - val_accuracy: 0.8346
Epoch 4/20
1719/1719 [==============================] - 15s 9ms/step - loss: 0.3626 - accuracy: 0.8662 - val_loss: 0.3684 - val_accuracy: 0.8662
Epoch 5/20
1719/1719 [==============================] - 15s 8ms/step - loss: 0.3546 - accuracy: 0.8676 - val_loss: 0.3409 - val_accuracy: 0.8764
Epoch 6/20
1719/1719 [==============================] - 13s 8ms/step - loss: 0.3526 - accuracy: 0.8683 - val_loss: 0.3670 - val_accuracy: 0.8572
Epoch 7/20
1719/1719 [==============================] - 12s 7ms/step - loss: 0.3480 - accuracy: 0.8705 - val_loss: 0.3717 - val_a

In each iteration the fit() method will call the object returned by max_norm() which will scale weights in return. In addition to rescaling the weights we can define different constraints, for instance, we can constrain bias terms by setting the bias_constrait argument. 

* A Dense layer usually has weights of shape [number of inputs, number of neurons], so using axis=0 means that the max-norm constraint will apply independently to each neuron’s weight vector. If you want to use max-norm with convolutional layers, make sure to set the max_norm() constraint’s axis argument appropriately (usually axis=[0, 1, 2]). - [Géron, A. (2019)](https://www.oreilly.com/library/view/hands-on-machine-learning/9781492032632/)

**Important documentations:**

https://www.tensorflow.org/api_docs/python/tf/keras/losses


https://www.tensorflow.org/api_docs/python/tf/keras/optimizers


https://www.tensorflow.org/api_docs/python/tf/keras/metrics

